plugins/source/azuredevops/codegen/recipes/base.go

```diff
@@ -0,0 +1,114 @@
+package recipes
+
+import (
+	"bytes"
+	"embed"
+	"fmt"
+	"go/format"
+	"os"
+	"path"
+	"reflect"
+	"runtime"
+	"text/template"
+
+	"github.com/cloudquery/plugin-sdk/codegen"
+	"github.com/cloudquery/plugin-sdk/schema"
+	"github.com/google/uuid"
+	"github.com/iancoleman/strcase"
+)
+
+//go:embed templates/*.go.tpl
+var templatesFS embed.FS
+
+type Resource struct {
+	Service    string
+	SubService string
+	Struct     interface{}
+	Table      *codegen.TableDefinition
+}
+
+func renderTemplate(name string, filePath string, data interface{}) error {
+	_, filename, _, ok := runtime.Caller(0)
+	if !ok {
+		return fmt.Errorf("failed to get caller information")
+	}
+	dir := path.Dir(filename)
+
+	tpl, err := template.New(name).Funcs(template.FuncMap{"ToCamel": strcase.ToCamel}).ParseFS(templatesFS, "templates/*.go.tpl")
+	if err != nil {
+		return fmt.Errorf("failed to parse azureDevops templates: %w", err)
+	}
+	tpl, err = tpl.ParseFS(codegen.TemplatesFS, "templates/*.go.tpl")
+	if err != nil {
+		return fmt.Errorf("failed to parse sdk template: %w", err)
+	}
+
+	var buff bytes.Buffer
+	if err := tpl.Execute(&buff, data); err != nil {
+		return fmt.Errorf("failed to execute template: %w", err)
+	}
+	outputPath := path.Join(dir, "../..", filePath)
+	outputDir := path.Dir(outputPath)
+	if err := os.MkdirAll(outputDir, os.ModePerm); err != nil {
+		return fmt.Errorf("failed to create directory %s: %w", outputDir, err)
+	}
+
+	content := buff.Bytes()
+	formattedContent, err := format.Source(buff.Bytes())
+	if err != nil {
+		fmt.Printf("failed to format source: %s: %v\n", outputPath, err)
+	} else {
+		content = formattedContent
+	}
+
+	if err := os.WriteFile(filePath, content, 0644); err != nil {
+		return fmt.Errorf("failed to write file %s: %w", outputPath, err)
+	}
+	return nil
+}
+
+func (resource *Resource) generate() error {
+	return renderTemplate("resource.go.tpl", path.Join("resources", "services", resource.Service, resource.SubService+".go"), resource)
+}
+
+func isUUID(fieldType reflect.Type) bool {
+	fieldKind := fieldType.Kind()
+
+	if fieldKind == reflect.Ptr {
+		return isUUID(fieldType.Elem())
+	}
+
+	return fieldType == reflect.TypeOf(uuid.UUID{})
+}
+
+func typeTransformer(field reflect.StructField) (schema.ValueType, error) {
+	if isUUID(field.Type) {
+		return schema.TypeUUID, nil
+	}
+
+	return codegen.DefaultTypeTransformer(field)
+}
+
+func (resource *Resource) Generate() error {
+	var err error
+	resource.Table, err = codegen.NewTableFromStruct(
+		fmt.Sprintf("azuredevops_%s_%s", resource.Service, resource.SubService),
```

Comment:
fmt would be overkill for 5 strings, maybe use `+` as bot concatenation of up to 5 strings it's optimized

Reply:
This is a pattern we use across all plugins. Can you explain `as bot concatenation of up to 5 strings it's optimized`?
I find it easier to understand how the final string looks like when using the `fmt.Sprintf`

Reply:
Go optimizes concat of up to 5 strings in compiler: `a + b + c + d + e` will be optimized, so it might be a bit faster way to do it.
As this is just codegen, no need for extra speed, though.